File: server/internal/domain/entities/character.go
package entities

import (
	"github.com/beka-birhanu/edit-space/server/internal/domain/types"

	"github.com/google/uuid"
)

type Character struct {
	Id         uuid.UUID      `json:"id" bun:",pk"`
	DocumentId uuid.UUID      `json:"documentId"`
	Position   types.Position `json:"position" bun:",array"`
	Value      string         `json:"value"`
}

func NewCharacter(id, documentId uuid.UUID, position types.Position, value string) Character {
	return Character{
		Id:         id,
		DocumentId: documentId,
		Position:   position,
		Value:      value,
	}
}

File: client/src/lib/crdt/character.ts

export type ChrPosition = number[];

class Character {
    id: string;
    documentId: string;
    position: ChrPosition;
    value: string;

    constructor(id: string, documentId: string, position: ChrPosition, value: string) {
        this.id = id;
        this.documentId = documentId;
        this.position = position;
        this.value = value;
    }

    static fromJSON(json: any): Character {
        return new Character(json.id, json.documentId, json.position, json.value);
    }

    toJSON(): any {
        return {
            id: this.id,
            position: this.position,
            value: this.value
        };
    }
}

export default Character;